doc/api/quic.md

```diff
@@ -0,0 +1,2276 @@
+    stream.setEncoding('utf8');
+    stream.on('data', console.log);
+    stream.on('end', () => console.log('stream ended'));
+  });
+});
+
+socket.on('listening', () => {
+  // The socket is listening for sessions!
+});
+```
+
+## QUIC Basics
+
+QUIC is a UDP-based network transport protocol that includes built-in security
+via TLS 1.3, flow control, error correction, connection migration,
+multiplexing, and more.
+
+Within the Node.js implementation of the QUIC protocol, there are three main
+components: the `QuicSocket`, the `QuicSession` and the `QuicStream`.
+
+### QuicSocket
+
+A `QuicSocket` encapsulates a binding to one or more local UDP ports. It is
+used to send data to, and receive data from, remote endpoints. Once created,
+a `QuicSocket` is associated with a local network address and IP port and can
+act as both a QUIC client and server simultaneously. User code at the
+JavaScript level interacts with the `QuicSocket` object to:
+
+* Query or modified the properties of the local UDP binding;
+* Create client `QuicSession` instances;
+* Wait for server `QuicSession` instances; or
+* Query activity statistics
+
+Unlike the `net.Socket` and `tls.TLSSocket`, a `QuicSocket` instance cannot be
+directly used by user code at the JavaScript level to send or receive data over
+the network.
+
+### Client and Server QuicSessions
+
+A `QuicSession` represents a logical connection between two QUIC endpoints (a
+client and a server). In the JavaScript API, each is represented by the
+`QuicClientSession` and `QuicServerSession` specializations.
+
+At any given time, a `QuicSession` exists is one of four possible states:
+
+* `Initial` - Entered as soon as the `QuicSession` is created.
+* `Handshake` - Entered as soon as the TLS 1.3 handshake between the client and
+  server begins. The handshake is always initiated by the client.
+* `Ready` - Entered as soon as the TLS 1.3 handshake completes. Once the
+  `QuicSession` enters the `Ready` state, it may be used to exchange
+  application data using `QuicStream` instances.
+* `Closed` - Entere as soon as the `QuicSession` connection has been
+  terminated.
+
+New instances of `QuicClientSession` are created using the `connect()`
+function on a `QuicSocket` as in the example below:
+
+```js
+const { createQuicSocket } = require('net');
+
+// Create a QuicSocket associated with localhost and port 1234
+const socket = createQuicSocket({ endpoint: { port: 1234 } });
+
+const client = socket.connect({
+  address: 'example.com',
+  port: 4567,
+  alpn: 'foo'
+});
+```
+
+As soon as the `QuicClientSession` is created, the `address` provided in
+the connect options will be resolved to an IP address (if necessary), and
+the TLS 1.3 handshake will begin. The `QuicClientSession` cannot be used
+to exchange application data until after the `'secure'` event has been
+emitted by the `QuicClientSession` object, signaling the completion of
+the TLS 1.3 handshake.
+
+```js
+client.on('secure', () => {
```

Comment:
Likely. Consider every part of the public facing JavaScript API bits on this a work in progress at this point. I'm open to changing any and every part of the JS API as necessary.

Semantically, in QUIC, the "connection" is considered to be established once the TLS handshake completes for a given side (note that these occur at different times between the client and server... that is, the server will consider the handshake completed at a different point in time than the client).

I originally went with `'secure'` here because of the distinction between 0RTT and normal QUIC traffic... where any stream data received during 0RTT cannot be considered fully secured. As I said tho, we can change the naming for sure.

Reply:
I don't want to obsess over the name, but they are hard to change later. Perhaps `'ready'` or `'handshaked'` would signal that this is not similar to the tls secure event, which means that the cert and other params have been checked.